File: examples/scripts/rust-orca/src/orca.rs
use ndarray::{arr1, Array1};

use crate::{
    halfplane::Halfplane,
    math::{
        angle2vec, angle_diff, arcsin, closest_point_on_line, cross, norm, normalize, vec2angle,
    },
    obstacle::Obstacle,
    participant::Participant,
};

const TAU: f64 = 4.0;

pub fn obstacle_collision(a: &Participant, obstacle: &Obstacle) -> (Array1<f64>, Array1<f64>) {
    let r_vec = normalize(&a.velocity) * (a.radius + a.safezone + obstacle.radius);
    let start = &obstacle.start - &(&a.position + &r_vec);
    let end = &obstacle.end - &(&a.position - &r_vec);

    let dist_vec =
        closest_point_on_line(&obstacle.start, &obstacle.end, &a.position, 0.0, 1.0) - &a.position;

    let u: Array1<f64>;
    let n: Array1<f64>;

    if norm(&dist_vec) < &a.radius + &a.safezone + &obstacle.radius {
        u = -&dist_vec - &a.velocity;
        n = normalize(&-dist_vec)
    } else {
        let start_tau = &start / TAU;
        let end_tau = &end / TAU;
        let closest = closest_point_on_line(&start_tau, &end_tau, &arr1(&[0.0, 0.0]), 0.0, 1.0);
        let w = closest;
        u = &w - &a.velocity;
        n = -normalize(&w);
    }

    return (u, n);
}

fn out_of_disk(
    disk_center: &Array1<f64>,
    disk_r: f64,
    velocity: &Array1<f64>,
) -> (Array1<f64>, Array1<f64>) {
    let rel_vec = velocity - disk_center;
    let w_length = norm(&rel_vec);
    let n = angle2vec(vec2angle(&rel_vec) + 10.0);
    let u = &n * (disk_r - w_length);
    return (u, n);
}

pub fn orca(a: &Participant, b: &Participant) -> (Array1<f64>, Array1<f64>) {
    let x = &b.position - &a.position;
    let r = &a.radius + &a.safezone + &b.radius + &b.safezone;
    let v = &a.velocity - &b.velocity;

    // check, if we are currently colliding
    if norm(&x) < r {
        return out_of_disk(&x, r, &v);
    }

    let disk_center = &x / TAU;
    let disk_r = r / TAU;
    let adjusted_disk_center = &disk_center * (1.0 - (r / norm(&x)).powi(2));
    // check, if we will collide with front disk
    if norm(&v) < norm(&adjusted_disk_center) && norm(&(&v - &adjusted_disk_center)) < r {
        return out_of_disk(&disk_center, disk_r, &v);
    }

    let position_angle = vec2angle(&x);
    let velocity_angle = vec2angle(&v);

    let difference_angle = angle_diff(position_angle, velocity_angle);

    let side_angle = arcsin(r, norm(&x));
    let right_side_angle = position_angle - side_angle;
    let left_side_angle = position_angle + side_angle;

    let right_side = angle2vec(right_side_angle);
    let left_side = angle2vec(left_side_angle);

    let left_point = closest_point_on_line(&arr1(&[0.0, 0.0]), &left_side, &v, 0.0, 1.0);
    let right_point = closest_point_on_line(&arr1(&[0.0, 0.0]), &right_side, &v, 0.0, 1.0);

    let left_u = &left_point - &v;
    let right_u = &right_point - &v;

    // let left_dist = dist(&left_point, &v);
    // let right_dist = dist(&right_point, &v);
    let left_dist = norm(&left_u);
    let right_dist = norm(&right_u);

    let u: Array1<f64>;
    if left_dist < right_dist {
        u = left_u;
    } else {
        u = right_u;
    }

    let mut n = normalize(&u);
    if difference_angle > side_angle {
        n *= -1.0
    }

    return (u, n);
}

pub fn halfplane_intersection(
    halfplanes_u: &Vec<Halfplane>,
    current_velocity: &Array1<f64>,
    optimal_point: &Array1<f64>,
) -> Option<Array1<f64>> {
    let mut halfplanes = Vec::new();
    for plane in halfplanes_u {
        halfplanes.push(Halfplane {
            u: current_velocity.clone() + plane.u.clone(),
            n: plane.n.clone(),
        });
    }
    let mut new_point = optimal_point.clone();

    for (i, plane) in halfplanes.iter().enumerate() {
        if (&new_point - &plane.u).dot(&plane.n) < 0.0 {
            let (left, right) = intersect_halfplane_with_other_halfplanes(plane, &halfplanes[..i]);
            if left.is_none() || right.is_none() {
                return None;
            }
            new_point = closest_point_on_line(
                &plane.u,
                &(&plane.u + &arr1(&[plane.n[1], -plane.n[0]])),
                optimal_point,
                left.unwrap(),
                right.unwrap(),
            )
        }
    }

    return Some(new_point);
}

fn intersect_halfplane_with_other_halfplanes(
    plane: &Halfplane,
    other_planes: &[Halfplane],
) -> (Option<f64>, Option<f64>) {
    let mut left = -f64::INFINITY;
    let mut right = f64::INFINITY;

    let direction = arr1(&[plane.n[1], -plane.n[0]]);

    for other_plane in other_planes {
        let other_dir = arr1(&[other_plane.n[1], -other_plane.n[0]]);
        let num = cross(&(&other_plane.u - &plane.u), &other_dir);
        let den = cross(&direction, &other_dir);

        if den == 0.0 {
            if num == 0.0 {
                return (None, None);
            }
            continue;
        }

        let t = &num / &den;
        if den > 0.0 {
            right = right.min(t);
        } else {
            left = left.max(t);
        }

        if left > right {
            return (None, None);
        }
    }

    return (Some(left), Some(right));
}

File: examples/scripts/python-orca/orca.py
import sys
from typing import List, Tuple

import numpy as np
from halfplane import Halfplane
from mathutils import (angle2vec, angle_diff, arcsin, closest_point_on_line,
                       dist, norm, normalize, vec2angle)
from obstacle import Obstacle
from participant import Participant

tau = 4


def obstacle_collision(a: Participant, obstacle: Obstacle) -> Tuple[np.ndarray, np.ndarray]:
    # TODO lome: rotate to take obstacle radius into account
    r_vec = normalize(a.velocity) * (a.radius + a.safezone + obstacle.radius)
    start = obstacle.start - (a.position + r_vec)
    end = obstacle.end - (a.position + r_vec)

    dist_vec = closest_point_on_line(
        obstacle.start, obstacle.end, a.position) - a.position
    # check, if we are colliding with the obstacle
    if norm(dist_vec) < a.radius + a.safezone + obstacle.radius:
        u = -dist_vec - a.velocity
        n = normalize(-dist_vec)
    else:
        # check, if we are colliding with VO of obstacle
        start_tau = start / tau
        end_tau = end / tau
        closest = closest_point_on_line(start_tau, end_tau, np.array([0, 0]))
        w = closest
        u = w - a.velocity
        n = -normalize(w)
    return u, n


def out_of_disk(disk_center, disk_r, v):
    # calculate vector & distance from center of disk to our velocity
    rel_vec = v - disk_center
    w_length = norm(rel_vec)
    # rotate vector by a certain degree, so we do not deadlock
    rel_vec = angle2vec(vec2angle(rel_vec) + 10)
    # calculate u (the velocity that will get us out of collision)
    u_vec = rel_vec * (disk_r - w_length)
    return u_vec, rel_vec


def orca(a: Participant, b: Participant) -> Tuple[np.ndarray, np.ndarray]:
    x = b.position - a.position
    r = a.radius + a.safezone + b.radius + b.safezone
    # r *= 2
    v = a.velocity - b.velocity

    if norm(x) < r:
        return out_of_disk(x, r, v)
    else:
        # calculate "tau-disk" center and radius
        disk_center = x / tau
        disk_r = r / tau

        # center of disk needs to be adjusted, sinec sides of the cone are not parallel
        adjusted_disk_center = disk_center * (1 - (r / norm(x))**2)

        # check, if we are colliding with the truncating disk
        if norm(v) < norm(adjusted_disk_center) and norm(v - adjusted_disk_center) < r:
            # we are colliding with truncating disk
            return out_of_disk(disk_center, disk_r, v)
        else:
            # calculate angles of relative position and velocity
            positionAngle = vec2angle(x)
            velocityAngle = vec2angle(v)

            # calculate angle between velocity and position vector
            differenceAngle = angle_diff(positionAngle, velocityAngle)

            # calculate angles of cone
            sideAngle = arcsin(r, norm(x))
            rightSideAngle = positionAngle - sideAngle
            leftSideAngle = positionAngle + sideAngle

            # calculate vectors for sides of code
            rightSide = angle2vec(rightSideAngle)
            leftSide = angle2vec(leftSideAngle)

            # calculate closest points on both side
            leftPoint = closest_point_on_line(
                np.array([0, 0]), leftSide, v)
            rightPoint = closest_point_on_line(
                np.array([0, 0]), rightSide, v)

            # calculate vectors for those closest points
            left_u = leftPoint - v
            right_u = rightPoint - v

            # calculate distance to closest points on lines
            leftDist = dist(leftPoint, v)
            rightDist = dist(rightPoint, v)

            # decide, which side is closer
            if leftDist < rightDist:
                u = left_u
            else:
                u = right_u

            # calculate n (the normal vector at the closest point pointing OUTWARDS)
            n = normalize(u)
            if differenceAngle > sideAngle:
                n *= -1
                u /= 2
            else:
                u *= 2

            # adjust u in relation to percentage of own radius
            # perc = 1
            # if norm(a.velocity) + norm(b.velocity) != 0:
            #     perc = 1 - (norm(a.velocity) /
            #                 (norm(a.velocity) + norm(b.velocity)))
            # perc = 1 - (a.radius / r)
            return u, n


def halfplane_intersection(halfplanes_u: List[Halfplane], current_velocity: np.ndarray, optimal_point: np.ndarray) -> np.ndarray:
    halfplanes = []
    for halfplane in halfplanes_u:
        halfplanes.append(
            Halfplane(current_velocity + halfplane.u, halfplane.n))
    new_point = optimal_point
    for i in range(len(halfplanes)):
        plane = halfplanes[i]
        # check, if we are out of the halfplane
        if np.dot(new_point - plane.u, plane.n) < 0:
            left, right = intersect_halfplane_with_other_halfplanes(
                plane, halfplanes[:i])
            # check, if feasible solution was found
            if left is None or right is None:
                return None
            new_point = closest_point_on_line(
                plane.u, plane.u + np.array([plane.n[1], -plane.n[0]]), optimal_point, left, right)
    return new_point


def intersect_halfplane_with_other_halfplanes(plane: Halfplane, other_planes: List[Halfplane]) -> Tuple[np.float64, np.float64]:
    left = float('-inf')
    right = float('inf')

    direction = np.array([plane.n[1], -plane.n[0]])

    # see https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect/565282#565282 for reference
    for other_plane in other_planes:
        other_dir = np.array([other_plane.n[1], -other_plane.n[0]])
        num = np.cross(other_plane.u - plane.u, other_dir)
        den = np.cross(direction, other_dir)

        if den == 0:
            # planes are parallel
            if num == 0:
                # planes are coincident
                return None, None
            else:
                # planes are parallel, but not coincident
                continue

        t = num / den
        if den > 0:
            # intersection is to the left of the line
            right = min(right, t)
        else:
            left = max(left, t)

        if left > right:
            return None, None
            # intersection is to the right of the line
    return left, right